File: monitor.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"sync"
	"time"
)

// CountryInfo 国家信息
type CountryInfo struct {
	ID      int    `json:"id"`
	CHN     string `json:"chn"`
	ENG     string `json:"eng"`
	Visible int    `json:"visible"`
}

// PriceData 价格数据
type PriceData struct {
	Cost          float64 `json:"cost"`
	Count         int     `json:"count"`
	PhysicalCount int     `json:"physicalCount"`
}

// Monitor 监控器
type Monitor struct {
	mu          sync.RWMutex
	config      *ConfigManager
	storage     *Storage
	notifier    *Notifier
	client      *http.Client
	countries   map[int]string
	services    map[string]string
	running     bool
	balance     float64
	lastCheck   string
	startTime   time.Time
	stopCh      chan struct{}
	eventCh     chan Event
	subscribers []chan Event
	subMu       sync.Mutex
}

// Event SSE事件
type Event struct {
	Type string      `json:"type"`
	Data interface{} `json:"data"`
}

// NewMonitor 创建监控器
func NewMonitor(config *ConfigManager, storage *Storage) *Monitor {
	return &Monitor{
		config:    config,
		storage:   storage,
		notifier:  NewNotifier(),
		client:    &http.Client{Timeout: 30 * time.Second},
		countries: make(map[int]string),
		services:  make(map[string]string),
		stopCh:    make(chan struct{}),
		eventCh:   make(chan Event, 100),
		startTime: time.Now(),
	}
}

// Init 初始化
func (m *Monitor) Init() {
	m.fetchCountries()
	m.fetchServices()
	m.storage.AddLog("info", "system", "系统启动", fmt.Sprintf("加载 %d 个国家", len(m.countries)))
}

// fetchCountries 获取国家列表
func (m *Monitor) fetchCountries() {
	cfg := m.config.Get()
	url := fmt.Sprintf("%s?action=getCountries&api_key=%s", cfg.BaseURL, cfg.APIKey)

	resp, err := m.client.Get(url)
	if err != nil {
		m.storage.AddLog("error", "system", "获取国家列表失败", err.Error())
		return
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var countries map[string]CountryInfo
	if err := json.Unmarshal(body, &countries); err != nil {
		return
	}

	m.mu.Lock()
	for _, info := range countries {
		if info.Visible == 1 {
			name := info.CHN
			if name == "" {
				name = info.ENG
			}
			m.countries[info.ID] = name
		}
	}
	m.mu.Unlock()
}

// fetchServices 获取服务列表
func (m *Monitor) fetchServices() {
	cfg := m.config.Get()
	url := fmt.Sprintf("%s?action=getServicesList&api_key=%s&lang=cn", cfg.BaseURL, cfg.APIKey)

	resp, err := m.client.Get(url)
	if err != nil {
		return
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var result struct {
		Services []struct {
			Code string `json:"code"`
			Name string `json:"name"`
		} `json:"services"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return
	}

	m.mu.Lock()
	for _, s := range result.Services {
		m.services[s.Code] = s.Name
	}
	m.mu.Unlock()
}

// GetCountryCount 获取国家数量
func (m *Monitor) GetCountryCount() int {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return len(m.countries)
}

// GetCountryName 获取国家名称
func (m *Monitor) GetCountryName(id int) string {
	m.mu.RLock()
	defer m.mu.RUnlock()
	if name, ok := m.countries[id]; ok {
		return name
	}
	return fmt.Sprintf("国家#%d", id)
}

// GetServices 获取服务列表
func (m *Monitor) GetServices() map[string]string {
	m.mu.RLock()
	defer m.mu.RUnlock()
	result := make(map[string]string)
	for k, v := range m.services {
		result[k] = v
	}
	return result
}

// FetchBalance 获取余额
func (m *Monitor) FetchBalance() (float64, error) {
	cfg := m.config.Get()
	url := fmt.Sprintf("%s?action=getBalance&api_key=%s", cfg.BaseURL, cfg.APIKey)

	resp, err := m.client.Get(url)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	text := string(body)

	var balance float64
	if _, err := fmt.Sscanf(text, "ACCESS_BALANCE:%f", &balance); err != nil {
		return 0, fmt.Errorf("unexpected response: %s", text)
	}

	m.mu.Lock()
	m.balance = balance
	m.mu.Unlock()

	// 检查余额告警
	cfgAlert := m.config.Get().BalanceAlert
	if cfgAlert.Enabled && balance < cfgAlert.Threshold {
		m.notifier.SendBalanceAlert(m.config.Get().Channels, balance)
	}

	return balance, nil
}

// FetchPrices 获取价格数据
func (m *Monitor) FetchPrices() (map[string]map[string]PriceData, error) {
	cfg := m.config.Get()
	url := fmt.Sprintf("%s?action=getPrices&api_key=%s", cfg.BaseURL, cfg.APIKey)

	resp, err := m.client.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var prices map[string]map[string]PriceData
	if err := json.Unmarshal(body, &prices); err != nil {
		return nil, err
	}

	return prices, nil
}

// RunCheck 执行一次检查
func (m *Monitor) RunCheck() ([]HitRecord, error) {
	start := time.Now()

	// 获取余额
	balance, err := m.FetchBalance()
	if err != nil {
		m.storage.AddLog("error", "check", "获取余额失败", err.Error())
		m.storage.AddCheck(CheckRecord{Status: "error", Error: err.Error(), Duration: time.Since(start).Milliseconds()})
		return nil, err
	}

	// 获取价格
	prices, err := m.FetchPrices()
	if err != nil {
		m.storage.AddLog("error", "check", "获取价格失败", err.Error())
		m.storage.AddCheck(CheckRecord{Status: "error", Error: err.Error(), Duration: time.Since(start).Milliseconds()})
		return nil, err
	}

	// 检查命中
	cfg := m.config.Get()
	var hits []HitRecord
	checkedAt := time.Now().Format("2006-01-02 15:04:05")

	for _, item := range cfg.Watchlist {
		if !item.Enabled {
			continue
		}

		countryStr := strconv.Itoa(item.Country)
		countryData, ok := prices[countryStr]
		if !ok {
			continue
		}

		serviceData, ok := countryData[item.Service]
		if !ok {
			continue
		}

		countryName := item.CountryName
		if countryName == "" {
			countryName = m.GetCountryName(item.Country)
		}

		hit := HitRecord{
			Timestamp:     checkedAt,
			Service:       item.Service,
			ServiceName:   item.ServiceName,
			Country:       item.Country,
			CountryName:   countryName,
			Cost:          serviceData.Cost,
			Count:         serviceData.Count,
			PhysicalCount: serviceData.PhysicalCount,
			MaxPrice:      item.MaxPrice,
		}

		isHit := serviceData.Cost <= item.MaxPrice && serviceData.Count > 0

		if isHit {
			hits = append(hits, hit)

			// 保存命中记录
			m.storage.AddHit(hit)

			// 记录消耗
			m.storage.AddCost(CostRecord{
				Service: item.ServiceName,
				Country: item.Country,
				Amount:  serviceData.Cost,
				Type:    "purchase",
			})
		}

		// 发送通知（命中或未命中都通知）
		go m.notifier.SendCheckNotification(cfg.Channels, hit, isHit)
	}

	// 记录日志
	if len(hits) > 0 {
		m.storage.AddLog("hit", "check", fmt.Sprintf("命中 %d 个号码", len(hits)), "")
	} else {
		m.storage.AddLog("info", "check", "本轮检查无命中", fmt.Sprintf("余额: $%.4f", balance))
	}

	// 记录检查记录
	m.storage.AddCheck(CheckRecord{
		Status:   "success",
		HitCount: len(hits),
		Duration: time.Since(start).Milliseconds(),
	})

	// 更新状态
	m.mu.Lock()
	m.lastCheck = time.Now().Format("2006-01-02 15:04:05")
	m.mu.Unlock()

	// 发送SSE事件
	m.sendEvent(Event{
		Type: "check",
		Data: map[string]interface{}{
			"balance":   balance,
			"hits":      hits,
			"time":      m.lastCheck,
			"hit_count": len(hits),
		},
	})

	return hits, nil
}

// GetStatus 获取监控状态
func (m *Monitor) GetStatus() MonitorStatus {
	m.mu.RLock()
	defer m.mu.RUnlock()
	cfg := m.config.Get()

	totalHits, totalChecks, errorCount := m.storage.GetStats()

	return MonitorStatus{
		Running:       m.running,
		Balance:       m.balance,
		CountryCount:  len(m.countries),
		LastCheck:     m.lastCheck,
		TotalHits:     totalHits,
		TotalChecks:   totalChecks,
		ErrorCount:    errorCount,
		WatchCount:    len(cfg.Watchlist),
		UptimeSeconds: int64(time.Since(m.startTime).Seconds()),
	}
}

// Start 启动监控
func (m *Monitor) Start() {
	m.mu.Lock()
	if m.running {
		m.mu.Unlock()
		return
	}
	m.running = true
	m.stopCh = make(chan struct{})
	m.mu.Unlock()

	m.storage.AddLog("info", "system", "监控已启动", "")
	m.sendEvent(Event{Type: "status", Data: "started"})

	go func() {
		// 立即执行一次
		if _, err := m.RunCheck(); err != nil {
			m.storage.AddLog("error", "system", "检查失败", err.Error())
		}

		interval := m.config.Get().IntervalSeconds
		if interval < 1 {
			interval = 60
		}
		ticker := time.NewTicker(time.Duration(interval) * time.Second)
		defer ticker.Stop()

		for {
			select {
			case <-m.stopCh:
				m.mu.Lock()
				m.running = false
				m.mu.Unlock()
				m.storage.AddLog("info", "system", "监控已停止", "")
				m.sendEvent(Event{Type: "status", Data: "stopped"})
				return
			case <-ticker.C:
				if _, err := m.RunCheck(); err != nil {
					m.storage.AddLog("error", "system", "检查失败", err.Error())
				}
			}
		}
	}()
}

// Stop 停止监控
func (m *Monitor) Stop() {
	m.mu.RLock()
	if !m.running {
		m.mu.RUnlock()
		return
	}
	m.mu.RUnlock()
	close(m.stopCh)
}

// IsRunning 是否运行中
func (m *Monitor) IsRunning() bool {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.running
}

// Subscribe 订阅事件
func (m *Monitor) Subscribe() chan Event {
	ch := make(chan Event, 100)
	m.subMu.Lock()
	m.subscribers = append(m.subscribers, ch)
	m.subMu.Unlock()
	return ch
}

// Unsubscribe 取消订阅
func (m *Monitor) Unsubscribe(ch chan Event) {
	m.subMu.Lock()
	defer m.subMu.Unlock()
	for i, sub := range m.subscribers {
		if sub == ch {
			m.subscribers = append(m.subscribers[:i], m.subscribers[i+1:]...)
			break
		}
	}
	close(ch)
}

// sendEvent 发送事件
func (m *Monitor) sendEvent(event Event) {
	m.subMu.Lock()
	defer m.subMu.Unlock()
	for _, ch := range m.subscribers {
		select {
		case ch <- event:
		default:
		}
	}
}

File: monitor.py
#!/usr/bin/env python3
"""HeroSMS 号码价格监控脚本。

监控指定服务在指定价格以下的可用号码数量。
"""

import hashlib
import hmac
import base64
import json
import subprocess
import time
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, Optional
from urllib.parse import urlencode


CONFIG_PATH = Path(__file__).parent / "config.json"


def load_config() -> Dict[str, Any]:
    """加载配置文件。"""
    with open(CONFIG_PATH, "r", encoding="utf-8") as f:
        return json.load(f)


def curl_get(url: str, timeout: int = 30) -> str:
    """使用 curl 发送 GET 请求。"""
    result = subprocess.run(
        ["curl", "-s", "-L", "--max-time", str(timeout), url],
        capture_output=True, text=True, timeout=timeout + 5,
    )
    if result.returncode != 0:
        raise RuntimeError(f"curl 失败: {result.stderr}")
    return result.stdout


def curl_post(url: str, payload: str, timeout: int = 10) -> str:
    """使用 curl 发送 POST 请求。"""
    result = subprocess.run(
        ["curl", "-s", "-L", "--max-time", str(timeout),
         "-X", "POST", "-H", "Content-Type: application/json",
         "-d", payload, url],
        capture_output=True, text=True, timeout=timeout + 5,
    )
    if result.returncode != 0:
        raise RuntimeError(f"curl 失败: {result.stderr}")
    return result.stdout


def fetch_country_names(api_key: str, base_url: str) -> Dict[int, str]:
    """从 API 获取国家 ID -> 中文名称映射。"""
    params = urlencode({"action": "getCountries", "api_key": api_key})
    url = f"{base_url}?{params}"
    names: Dict[int, str] = {}
    try:
        data = json.loads(curl_get(url))
        for cid, info in data.items():
            names[int(cid)] = info.get("chn", info.get("eng", f"国家#{cid}"))
    except Exception:
        pass
    return names


def fetch_offers(api_key: str, base_url: str) -> Dict[str, Any]:
    """获取所有服务的报价数据。"""
    params = urlencode({"action": "getPrices", "api_key": api_key})
    url = f"{base_url}?{params}"
    return json.loads(curl_get(url))


def fetch_balance(api_key: str, base_url: str) -> Optional[float]:
    """查询账户余额。"""
    params = urlencode({"action": "getBalance", "api_key": api_key})
    url = f"{base_url}?{params}"
    try:
        text = curl_get(url, timeout=15)
        if text.startswith("ACCESS_BALANCE:"):
            return float(text.split(":")[1])
        return None
    except Exception:
        return None


def check_hits(
    offers: Dict[str, Any],
    watchlist: List[Dict[str, Any]],
    country_names: Dict[int, str],
) -> List[Dict[str, Any]]:
    """检查监控列表，返回命中项。

    offers 格式: {country_id: {service_id: {cost, count, physicalCount}}}
    """
    hits: List[Dict[str, Any]] = []

    for item in watchlist:
        service = item["service"]
        country = str(item["country"])
        max_price = item["max_price"]
        service_name = item.get("service_name", service)
        country_name = item.get("country_name", country_names.get(int(country), f"国家#{country}"))

        # 获取该国家该服务的数据
        country_data = offers.get(country)
        if not country_data:
            continue

        service_data = country_data.get(service)
        if not service_data:
            continue

        cost = service_data.get("cost", 0)
        count = service_data.get("count", 0)
        physical_count = service_data.get("physicalCount", 0)

        # 检查是否低于目标价格且有号码
        if cost <= max_price and count > 0:
            hits.append({
                "service": service,
                "service_name": service_name,
                "country": int(country),
                "country_name": country_name,
                "cost": cost,
                "count": count,
                "physical_count": physical_count,
                "max_price": max_price,
            })

    return hits


def format_hit(hit: Dict[str, Any], checked_at: str) -> List[str]:
    """格式化单条命中消息。"""
    return [
        f"🎯 {hit['service_name']} 号码命中",
        "",
        hit["country_name"],
        f"💰 当前价格：${hit['cost']}",
        f"📦 可用数量：{hit['count']}",
        f"🎚 目标价格：≤ ${hit['max_price']}",
        "",
        f"✅ 可用号码数：{hit['count']}（物理卡 {hit['physical_count']}）",
        f"🕒 检查时间：{checked_at}",
    ]


def format_no_hit(watchlist: List[Dict[str, Any]], checked_at: str) -> List[str]:
    """格式化无命中消息。"""
    lines = ["📭 本轮检查无命中", ""]
    for item in watchlist:
        lines.append(f"  {item.get('service_name', item['service'])} @ {item.get('country_name', str(item['country']))}：≤ ${item['max_price']}")
    lines.append("")
    lines.append(f"🕒 检查时间：{checked_at}")
    return lines


def print_lines(lines: List[str]) -> None:
    """打印消息行。"""
    for line in lines:
        print(line)


def gen_feishu_sign(secret: str, timestamp: int) -> str:
    """生成飞书 Webhook 签名。"""
    string_to_sign = f"{timestamp}\n{secret}"
    hmac_code = hmac.new(string_to_sign.encode("utf-8"), digestmod=hashlib.sha256).digest()
    return base64.b64encode(hmac_code).decode("utf-8")


def send_feishu_webhook(webhook_url: str, hits: List[Dict[str, Any]], checked_at: str, sign: Optional[str] = None) -> None:
    """通过飞书 Webhook 发送命中通知。"""
    for hit in hits:
        content = (
            f"🎯 {hit['service_name']} 号码命中\n"
            f"国家：{hit['country_name']}\n"
            f"💰 当前价格：${hit['cost']}\n"
            f"📦 可用数量：{hit['count']}\n"
            f"🎚 目标价格：≤ ${hit['max_price']}\n"
            f"✅ 可用号码数：{hit['count']}（物理卡 {hit['physical_count']}）\n"
            f"🕒 检查时间：{checked_at}"
        )
        body: Dict[str, Any] = {
            "msg_type": "text",
            "content": {"text": content},
        }

        # 加签
        if sign:
            timestamp = int(time.time())
            body["timestamp"] = str(timestamp)
            body["sign"] = gen_feishu_sign(sign, timestamp)

        payload = json.dumps(body)
        try:
            result = curl_post(webhook_url, payload)
            print(f"📨 飞书通知已发送：{result}")
        except Exception as e:
            print(f"❌ 飞书通知失败：{e}")


def run_once(config: Dict[str, Any], country_names: Dict[int, str]) -> List[Dict[str, Any]]:
    """执行一次检查，返回命中列表。"""
    api_key = config["api_key"]
    base_url = config.get("base_url", "https://hero-sms.com/stubs/handler_api.php")
    watchlist = config["watchlist"]
    checked_at = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

    if not api_key:
        print("❌ 错误：未配置 api_key，请编辑 config.json")
        return []

    # 查询余额
    balance = fetch_balance(api_key, base_url)
    if balance is not None:
        print(f"💰 账户余额：${balance}")

    # 获取报价
    offers = fetch_offers(api_key, base_url)

    # 检查命中
    hits = check_hits(offers, watchlist, country_names)

    if hits:
        for hit in hits:
            print_lines(format_hit(hit, checked_at))
            print()

        # 飞书通知
        webhook = config.get("webhook", {})
        if webhook.get("enabled") and webhook.get("url"):
            send_feishu_webhook(webhook["url"], hits, checked_at, webhook.get("sign"))
    else:
        print_lines(format_no_hit(watchlist, checked_at))
        print()

    return hits


def main() -> None:
    """主函数：循环监控。"""
    config = load_config()
    api_key = config["api_key"]
    base_url = config.get("base_url", "https://hero-sms.com/stubs/handler_api.php")
    interval = config.get("interval_seconds", 60)

    # 启动时获取国家名称映射
    country_names = fetch_country_names(api_key, base_url)

    print("=" * 50)
    print("  HeroSMS 号码价格监控")
    print("=" * 50)
    print(f"监控间隔：{interval} 秒")
    print(f"监控项数：{len(config['watchlist'])}")
    print(f"国家库：{len(country_names)} 个国家")
    print("=" * 50)
    print()

    while True:
        try:
            run_once(config, country_names)
        except KeyboardInterrupt:
            print("\n监控已停止。")
            break
        except Exception as e:
            print(f"❌ 检查出错：{e}")

        try:
            time.sleep(interval)
        except KeyboardInterrupt:
            print("\n监控已停止。")
            break


if __name__ == "__main__":
    main()
